cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/etl/planner/ControlDagTest.java

```diff
@@ -0,0 +1,318 @@
+    cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n1", "n4"),
+        new Connection("n2", "n5"),
+        new Connection("n3", "n5"),
+        new Connection("n4", "n5"),
+        new Connection("n5", "n6"),
+        new Connection("n5", "n7"),
+        new Connection("n6", "n8"),
+        new Connection("n7", "n8")));
+    cdag.flatten();
+    expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n1", "n4"),
+        new Connection("n2", "n5"),
+        new Connection("n3", "n5"),
+        new Connection("n4", "n5"),
+        new Connection("n5", "n6"),
+        new Connection("n5", "n7"),
+        new Connection("n6", "n8"),
+        new Connection("n7", "n8")));
+    Assert.assertEquals(expected, cdag);
+  }
+
+  @Test
+  public void testMultiSourceFlatten() {
+    /*
+        n1 --|
+             |--> n3
+        n2 --|
+     */
+    ControlDag cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n3"),
+        new Connection("n2", "n3")));
+    cdag.flatten();
+    ControlDag expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1.n2", "n1"),
+        new Connection("n1.n2", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n2", "n3")));
+    Assert.assertEquals(expected, cdag);
+  }
+
+  @Test
+  public void testFlatten() {
+    /*
+                      |--> n3
+            |--> n2 --|
+            |         |--> n4
+       n1 --|
+            |
+            |--> n5
+     */
+    ControlDag cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n5"),
+        new Connection("n2", "n3"),
+        new Connection("n2", "n4")));
+    cdag.flatten();
+    /*
+            |--> n2 --|             |--> n3 --|
+       n1 --|         |--> n2.n5 -->|         |--> n3.n4
+            |--> n5 --|             |--> n4 --|
+     */
+    ControlDag expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n5"),
+        new Connection("n2", "n2.n5"),
+        new Connection("n5", "n2.n5"),
+        new Connection("n2.n5", "n3"),
+        new Connection("n2.n5", "n4"),
+        new Connection("n3", "n3.n4"),
+        new Connection("n4", "n3.n4")));
+    Assert.assertEquals(expected, cdag);
+
+    /*
+                      |--> n3
+            |--> n2 --|
+            |         |--> n4
+       n1 --|              |
+            |              v
+            |--> n5 -----> n6
+     */
+    cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n5"),
+        new Connection("n2", "n3"),
+        new Connection("n2", "n4"),
+        new Connection("n4", "n6"),
+        new Connection("n5", "n6")));
+    cdag.flatten();
```

Comment:
true, the dag ends up more complicated.  But configuring the workflow becomes significantly easier if we don't have to worry about nested forks, especially since the fork() method returns a different object than what you call it on, making recursion... difficult.